### ISSUE TITLE
Implement zfs recv buffer

### ISSUE THREAD

**Issue by ryao** [zfsonlinux/zfs#1161]
UNIX pipes usually have a 64KB buffer size, which is too small to buffer a ZFS transaction. The consequence is that a zfs send | ... | zfs recv operation will typically alternate between sending and receiving, which is suboptimal. A program called `mbuffer` had been suggested by various ZFS users as a workaround for this. `mbuffer` provides user adjustable buffer that is 2MB in size by default, which is generally sufficient to avoid suboptimal behavior in practice.

I encountered an issue where a zfs send stopped prematurely when I was using `mbuffer`, which has caused me to question its reliability. It would be ideal to integrate this functionality into the `zfs recv` command to ensure that buffering is done in a consistent manner. This would have the additional benefit of ensuring that users do not accidentally place mbuffer on the `zfs send` side of a SSH tunnel, which would reduce the benefit of a buffer.

**Comment by lintonv** [comment]
@olw2005 could you share what your "optimum" was for your setup?

**Comment by olw2005** [comment]
@lintonv

We tinkered with lz4demo:         http://code.google.com/p/lz4/
and mbuffer:     http://www.maier-komor.de/mbuffer.html

locally compiled along with a modified version of the “zfs-replicate” shell script from here:

#
# Author: kattunga
# Date: August 11, 2012
# Version: 2.0
#
# http://linuxzfs.blogspot.com/2012/08/zfs-replication-script.html
# https://github.com/kattunga/zfs-scripts.git
#
# Credits:
#    Mike La Spina for the original concept and script http://blog.laspina.ca/
#
# Function:
#    Provides snapshot and send process which replicates a ZFS dataset from a source to target server.
#    Maintains a runing snapshot archive for X time
#

The modified zfs send [and ssh -> zfs receive on the other end] looked like this:

zfs send $VERBOSE $DEDUP -R $last_snap_source | lz4demo stdin stdout 2> /dev/null | mbuffer -q -m 512M 2> /dev/null | ssh -c aes128-cbc $TGT_HOST $TGT_PORT "lz4demo -d stdin stdout 2> /dev/null | zfs recv $VERBOSE -F $TGT_PATH" 2> $0.err

But in the end, the above did not significantly outperform straight ssh (with aes128-cbc encryption).

YMMV.

**Comment by eborisch** [comment]
FWIW, lz4 (or lz4c) is available on many distros in some form, so you likely don't need to roll your own anymore.

We've been happy:
 zfs send [args ...] | lz4c | ssh remote_host "mbuffer [args to rate limit /buffer] | lz4c -d | zfs recv [args]"

If you aren't rate limiting, mbuffer may still allow the send to finish faster if you are sending small incrementals, especially with multiple small (hourly, for example) snapshots. You can also set your ssh cipher to arcfour to lower ssh's cpu load if you don't need military grade encryption

**Comment by lintonv** [comment]
@olw2005 @eborisch Thank you both. I'll do some testing and post what I find.

**Comment by olw2005** [comment]
@eborisch @lintonv  If you have AES-NI instruction sets (i.e. a newer cpu) the speed for aes-128-cbc is pretty decent.  Fast enough for my use case, anyway.  RH has a web page with test cmds for ssh here:
https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/sect-Security_Guide-Encryption-OpenSSL_Intel_AES-NI_Engine.html

Good luck!

--------------------------------------------------------------------------------------------------------------------------------------------------------------
This e-mail and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are
addressed. If you have received this e-mail in error please notify the sender immediately and delete this e-mail from your system.
Please note that any views or opinions presented in this e-mail are solely those of the author and do not necessarily represent those
of Ricardo (save for reports and other documentation formally approved and signed for release to the intended recipient). Only Directors
are authorised to enter into legally binding obligations on behalf of Ricardo. Ricardo may monitor outgoing and incoming e-mails and
other telecommunications systems. By replying to this e-mail you give consent to such monitoring. The recipient should check e-mail and
any attachments for the presence of viruses. Ricardo accepts no liability for any damage caused by any virus transmitted by this e-mail.
"Ricardo" means Ricardo plc and its subsidiary companies.
Ricardo plc is a public limited company registered in England with registered number 00222915.
The registered office of Ricardo plc is Shoreham Technical Centre, Shoreham-by Sea, West Sussex, BN43 5FG.
--------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by eborisch** [comment]
And as I've mentioned elsewhere before (#3010) I would suggest avoiding '-F' on the recv if at all possible.

**Comment by lintonv** [comment]
@eborisch @olw2005 

I tried mbuffer and using the lz4c compression, in the ways you both suggested above. 

But, I still see bad send performance initially. I expect 120 MB/sec but I only get 12 MB/sec. 

Let me explain:
1. Initial send of a 1 Gig FS goes at 12 MB/sec
2. After that transfer, delete the FS on the RECEIVING end and re-transmit. At this point, I get the full bandwidth of 120 MB/sec,

This shows me that there is some caching (probably ZIL?) which is why the second send is much faster. But the initial send (with no cache?) is extremely slow.

**Comment by olw2005** [comment]
@lintonv  You might try directing the send into /dev/null to eliminate other variables.  It sounds like your disk may be the bottleneck, in which case buffering / compressing won’t help.

**Comment by lintonv** [comment]
@olw2005 It does not appear to be the disks. I use enterprise grade SSDs whose bandwidth and speed are very high. That is not the bottleneck.

I did some additional tests and what I discovered was a 'queue depth' of 1. ZFS send appears to be a highly serial operation. I am going to look at the code and see if parallelism is possible. 

Any other insights on how this can be done in a parallel fashion?

**Comment by olw2005** [comment]
@lintonv I’ll leave the code questions for others to answer.  I’m a sysadmin not a programmer, Jim.  =)

However I will note that in our usage, an unconstrained (redirected to /dev/null for example) “full” zfs send of an un-cached but relatively un-fragmented filesystem easily pegs the 6gbps sas controller.  We typically net around 500 - 600 MB/s @ around 4k – 5k iops which is about right given (raw speed) * compression / (raidz2 overhead).  In practice we get around 250-300 MB/s dumping a zfs send out across the 10gbit lan to LTO5 tape on a backup server.  (I believe in that case it’s largely constrained by the tape speed.)

Bottom line, I don’t think there is anything “wrong” with the zfs send code.

**Comment by lintonv** [comment]
I did not mean to hijack this thread. I apologize. As my issue is on the ZFS send side and not on the zfs recv side, I will stop here. 

Just FYI, I did some improvements by using a larger record size (I was using 4K) and turning primarycache to 'all'. But still nothing significant.

**Comment by olw2005** [comment]
@lintonv  On that note, I was going to mention in my last post but forgot.  You should redirect the performance questions to the zfs discussion list (see the zfsonlinux.org website).  You’ll get more advice there.  (In fact, if you search it you’ll probably find the question of zfs send/recv performance has come up before.  Repeatedly.)

As for block size, this may be out-of-date but at the time we implemented (circa v0.6.1) the 128k block size worked a lot better for our use case (zvols shared with iscsi to vmware).  I tested a range of block sizes and there was noticeable performance degradation at the smallest block sizes (4k and 8k in particular).  Again, take it with a grain of salt as that was about 3 years ago and the zfs code has changed a lot since then.

**Comment by ryao** [comment]
I noticed when reviewing documentation that it is possible for userspace to use `fctntl(fd, F_SETPIPE_SZ, size)` to change the kernel pipe buffer size on Linux to increase the pipe size up to the value specified in /proc/sys/fs/pipe-max-size. We can use `fstat` to check if the fd is of type  S_IFIFO so that we only do this on actual pipes. I thought of it when working on something else, so I am making a note here should someone else want to do it before I find time. This should be trivial to achieve.

**Comment by ryao** [comment]
Those additional pushes were just for changes to the commit message. Anyway, the triviality of this piqued my interest, so I implemented it, compiled  it and verified with strace that the right syscalls were being done on a simple test case. Someone else will need to verify that it actually provides a benefit.

**Comment by ryao** [comment]
Would someone who benefits from mbuffer mind doing a benchmark of ryao/zfs@3530cf2bc933e0fc6a035026934af8976542bfff against the unpatched userland binaries with and without mbuffer? It should outperform mbuffer in situations where a single core cannot keep up with the two additional copies using mbuffer requires while eliminating the need for it.

**Comment by ryao** [comment]
5c3f61eb498e8124858b1369096bf64b86a938e7 closed this.
